imagenet/main.py

```diff
@@ -83,12 +94,49 @@ def main():
         warnings.warn('You have chosen a specific GPU. This will completely '
                       'disable data parallelism.')
 
-    args.distributed = args.world_size > 1
+    args.distributed = args.world_size > 1 or args.multiprocessing_distributed
+
+    ngpus_per_node = torch.cuda.device_count()
+    if args.multiprocessing_distributed:
+        # python version check
+        if not ((sys.version_info[0] == 3 and sys.version_info[1] >= 4) or
+                sys.version_info[0] > 3):
+            raise RuntimeError("Requires python 3.4 or higher to use "
+                               "torch.multiprocessing.spawn helper to launch "
+                               "multiple processes for distributed training. "
+                               "If you have a lower version of Python, use "
+                               "torch.distributed.launch instead.")
+        # Since we have ngpus_per_node processes per node, the total world_size
+        # needs to be adjusted accordingly
+        args.world_size = ngpus_per_node * args.world_size
+        # Use torch.multiprocessing.spawn to launch distributed processes: the
+        # main_worker process function
+        mp.spawn(
+            main_worker,
+            nprocs=ngpus_per_node,
+            args=(ngpus_per_node, args)
+        )
+    else:
+        # Simply call main_worker function
+        main_worker(args.gpu, ngpus_per_node, args)
 
-    if args.distributed:
-        dist.init_process_group(backend=args.dist_backend, init_method=args.dist_url,
-                                world_size=args.world_size)
 
+def main_worker(gpu, ngpus_per_node, args):
+    global best_acc1
+    args.gpu = gpu
+
+    if args.gpu is not None:
+        print("Use GPU: {} for training".format(args.gpu))
+
+    if args.distributed:
+        if args.multiprocessing_distributed:
+            # For multiprocessing distributed training, rank needs to be the
+            # global rank among all the processes
+            args.rank = args.rank * ngpus_per_node + gpu
+        dist.init_process_group(backend=args.dist_backend,
```

Comment:
nit: our line limits are 120 chars. why break the lines before we hit the limit.

Reply:
addressed